README.md
# unbound

Tested with Travis CI

[![Build Status](https://travis-ci.org/bodgit/puppet-unbound.svg?branch=master)](https://travis-ci.org/bodgit/puppet-unbound)
[![Coverage Status](https://coveralls.io/repos/bodgit/puppet-unbound/badge.svg?branch=master&service=github)](https://coveralls.io/github/bodgit/puppet-unbound?branch=master)
[![Puppet Forge](http://img.shields.io/puppetforge/v/bodgit/unbound.svg)](https://forge.puppetlabs.com/bodgit/unbound)
[![Dependency Status](https://gemnasium.com/bodgit/puppet-unbound.svg)](https://gemnasium.com/bodgit/puppet-unbound)

#### Table of Contents

1. [Description](#description)
2. [Setup - The basics of getting started with unbound](#setup)
    * [Beginning with unbound](#beginning-with-unbound)
3. [Usage - Configuration options and additional functionality](#usage)
4. [Reference - An under-the-hood peek at what the module is doing and how](#reference)
5. [Limitations - OS compatibility, etc.](#limitations)
6. [Development - Guide for contributing to the module](#development)

## Description

This module manages Unbound.

RHEL/CentOS and OpenBSD are supported using Puppet 4.6.0 or later.

## Setup

### Beginning with unbound

In the simplest case, configure Unbound listening on all interfaces:

```puppet
include ::unbound
```

## Usage

Update the above example to specify the forwarders to use and allow access
from the local network, also restrict outbound ports used to avoid anything
below 32768 and above 65000:

```puppet
class { '::unbound':
  access_control => [
    ["${::network}/${netmask_to_masklen($::netmask)}", 'allow'],
    ["${::network6}/${netmask_to_masklen($::netmask6)}", 'allow'],
  ],
  outgoing_port  => [
    ['avoid', 0, 32767],
    ['avoid', 65001, 65535],
  ],
}

::unbound::forward { '.':
  forward_addr => [
    '8.8.8.8',
    '8.8.4.4',
    '2001:4860:4860::8844',
    '2001:4860:4860::8888',
  ],
}
```

Configure NSD listening on the primary interface as a slave for a single
zone protected with a given TSIG key and configure Unbound listening on
the loopback interface only with a stub zone pointing to NSD:

```puppet
class { '::nsd':
  ip_address => [
    $::ipaddress,
    $::ipaddress6,
  ],
}

::nsd::key { 'example.':
  algorithm => 'hmac-sha256',
  secret    => '6z+8iKRIQrwN43TFfO/Rf2NHzpHIFVi6PsJ7dDESclc=',
}

::nsd::zone { 'example.com.':
  allow_notify => [
    ['192.0.2.1', 'example.'],
  ],
  request_xfr  => [
    ['AXFR', '192.0.2.1', 'example.'],
  ],
}

class { '::unbound':
  interface => [
    $::ipaddress_lo,
    $::ipaddress6_lo,
  ],
}

::unbound::stub { 'example.com.':
  stub_addr => [
    $::ipaddress,
    $::ipaddress6,
  ],
}
```

If NSD is not required then local data can be used instead:

```puppet
class { '::unbound':
  interface  => [
    $::ipaddress_lo,
    $::ipaddress6_lo,
  ],
  local_zone => [
    ['example.com.', 'static'],
  ],
  local_data => [
    {
      'name'     => 'example.com.'
      'ttl'      => 86400,
      'class'    => 'IN',
      'type'     => 'NS',
      'hostname' => 'ns1.example.com.',
    },
    {
      'name'     => 'example.com.'
      'ttl'      => 86400,
      'class'    => 'IN',
      'type'     => 'SOA',
      'primary'  => 'ns1.example.com.',
      'email'    => 'hostmaster.example.com.',
      'serial'   => 1,
      'refresh'  => 10800,
      'retry'    => 15,
      'expire'   => 604800,
      'negative' => 10800,
    },
    {
      'name'  => 'ns1.example.com.',
      'ttl'   => 86400,
      'class' => 'IN',
      'type'  => 'A',
      'ip'    => '192.0.2.1',
    },
  ],
}
```

## Reference

The reference documentation is generated with
[puppet-strings](https://github.com/puppetlabs/puppet-strings) and the latest
version of the documentation is hosted at
[https://bodgit.github.io/puppet-unbound/](https://bodgit.github.io/puppet-unbound/).

## Limitations

This module has been built on and tested against Puppet 4.6.0 and higher.

The module has been tested on:

* RedHat Enterprise Linux 6/7
* OpenBSD 6.0/6.1

## Development

The module has both [rspec-puppet](http://rspec-puppet.com) and
[beaker-rspec](https://github.com/puppetlabs/beaker-rspec) tests. Run them
with:

```
$ bundle exec rake test
$ PUPPET_INSTALL_TYPE=agent PUPPET_INSTALL_VERSION=x.y.z bundle exec rake beaker:<nodeset>
```

Please log issues or pull requests at
[github](https://github.com/bodgit/puppet-unbound).


CHANGELOG.md
# Change Log

## [v2.0.0](https://github.com/bodgit/puppet-unbound/tree/v2.0.0) (2017-10-15)
[Full Changelog](https://github.com/bodgit/puppet-unbound/compare/v1.0.0...v2.0.0)

**Implemented enhancements:**

- Puppet 4 [\#1](https://github.com/bodgit/puppet-unbound/issues/1)

**Merged pull requests:**

- V2 [\#2](https://github.com/bodgit/puppet-unbound/pull/2) ([bodgit](https://github.com/bodgit))

## [v1.0.0](https://github.com/bodgit/puppet-unbound/tree/v1.0.0) (2016-07-30)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

spec/aliases/acl_spec.rb
require 'spec_helper'

if Puppet::Util::Package.versioncmp(Puppet.version, '4.4.0') >= 0
  describe 'test::acl', type: :class do
    describe 'accepts ACLs' do
      [
        ['127.0.0.1', 'allow'],
        ['::1', 'allow_snoop'],
        ['192.0.2.1', 'deny'],
        ['2001:db8::1', 'deny_non_local'],
        ['192.0.2.0/24', 'refuse'],
        ['2001:db8::/32', 'refuse_non_local'],
      ].each do |value|
        describe value.inspect do
          let(:params) {{ value: value }}
          it { is_expected.to compile }
        end
      end
    end
    describe 'rejects other values' do
      [
        1,
        ['invalid', 'allow'],
        ['127.0.0.1', 'invalid'],
      ].each do |value|
        describe value.inspect do
          let(:params) {{ value: value }}
          it { is_expected.to compile.and_raise_error(/parameter 'value' /) }
        end
      end
    end
  end
end


spec/aliases/port_spec.rb
require 'spec_helper'

if Puppet::Util::Package.versioncmp(Puppet.version, '4.4.0') >= 0
  describe 'test::port', type: :class do
    describe 'accepts ports' do
      [
        ['permit', 1024],
        ['permit', 1024, 65535],
        ['avoid', 1024],
        ['avoid', 1024, 65535],
      ].each do |value|
        describe value.inspect do
          let(:params) {{ value: value }}
          it { is_expected.to compile }
        end
      end
    end
    describe 'rejects other values' do
      [
        ['invalid', 65536],
        ['invalid', 1024, 65536],
      ].each do |value|
        describe value.inspect do
          let(:params) {{ value: value }}
          it { is_expected.to compile.and_raise_error(/parameter 'value' /) }
        end
      end
    end
  end
end


spec/aliases/size_spec.rb
require 'spec_helper'

if Puppet::Util::Package.versioncmp(Puppet.version, '4.4.0') >= 0
  describe 'test::size', type: :class do
    describe 'accepts sizes' do
      [
        0,
        '0',
        '1k',
        '1m',
        '1g',
      ].each do |value|
        describe value.inspect do
          let(:params) {{ value: value }}
          it { is_expected.to compile }
        end
      end
    end
    describe 'rejects other values' do
      [
        -1,
        '-1',
      ].each do |value|
        describe value.inspect do
          let(:params) {{ value: value }}
          it { is_expected.to compile.and_raise_error(/parameter 'value' /) }
        end
      end
    end
  end
end


lib/puppet/parser/functions/validate_unbound_port.rb
#
# validate_unbound_port.rb
#

module Puppet::Parser::Functions
  newfunction(:validate_unbound_port, :doc => <<-EOS
    Validate that all passed values are all valid port specifications.
    Each specification must begin with either `permit` or `avoid` followed
    by either a single port or a range joined with `-`.

    Example:

      validate_unbound_port(['permit 1000', 'avoid 500-510'])
    EOS
  ) do |arguments|

    raise Puppet::ParserError, 'validate_unbound_port(): Wrong number of ' +
      "arguments given (#{arguments.size} for 1)" if arguments.size != 1

    item = arguments[0]

    unless item.is_a?(Array)
      item = [item]
    end

    if item.size == 0
      raise Puppet::ParseError, 'validate_unbound_port(): Requires an array ' +
        'with at least 1 element'
    end

    item.each do |i|
      unless i.is_a?(String)
        raise Puppet::ParseError, 'validate_unbound_port(): Requires either ' +
          'an array or string to work with'
      end

      if i =~ /^ (?:permit|avoid) \s+ (\d+) (?:-(\d+))? $/x
        from = $1.to_i
        to   = $2 ? $2.to_i : nil
        if from > 65535 or (to and (to > 65535 or to <= from))
          raise Puppet::ParseError, 'validate_unbound_port(): ' +
            "'#{i.inspect}' is not a valid port"
        end
      else
        raise Puppet::ParseError, 'validate_unbound_port(): ' +
          "'#{i.inspect}' is not a valid port"
      end
    end
  end
end


lib/puppet/parser/functions/validate_unbound_ratelimit.rb
#
# validate_unbound_ratelimit.rb
#

module Puppet::Parser::Functions
  newfunction(:validate_unbound_ratelimit, :doc => <<-EOS
    Validate that all passed values are all valid rate limits. Each rate
    limit should be a domain name followed by a rate specified as an
    integer.

    Example:

      validate_unbound_ratelimit(['example.com 1000', 'example.org 2000'])
    EOS
  ) do |arguments|

    raise Puppet::ParserError, 'validate_unbound_ratelimit(): Wrong number of ' +
      "arguments given (#{arguments.size} for 1)" if arguments.size != 1

    item = arguments[0]

    unless item.is_a?(Array)
      item = [item]
    end

    if item.size == 0
      raise Puppet::ParseError, 'validate_unbound_ratelimit(): Requires an array ' +
        'with at least 1 element'
    end

    item.each do |i|
      unless i.is_a?(String)
        raise Puppet::ParseError, 'validate_unbound_ratelimit(): Requires either ' +
          'an array or string to work with'
      end

      if i =~ /^ (\S+) \s+ \d+ $/x
        if not function_is_domain_name([$1])
          raise Puppet::ParseError, 'validate_unbound_ratelimit(): ' +
            "'#{i.inspect}' is not a valid ratelimit"
        end
      else
        raise Puppet::ParseError, 'validate_unbound_ratelimit(): ' +
          "'#{i.inspect}' is not a valid ratelimit"
      end
    end
  end
end


lib/puppet/parser/functions/validate_unbound_acl.rb
#
# validate_unbound_acl.rb
#

require 'ipaddr'

# Patch IP class so the netmask is accessible
class IPAddr
  attr_reader :mask_addr
end

module Puppet::Parser::Functions
  newfunction(:validate_unbound_acl, :doc => <<-EOS
    Validate that all passed values are all valid IP addresses. A second
    argument passes one or more fixed strings that are valid values for
    actions that may follow the address. A final third parameter controls
    if the IP address accepts single addresses or CIDR netblocks.

    Example:

      validate_unbound_acl(['1.2.3.4', '5.6.7.8@5353'])
      validate_unbound_acl(['1.2.3.0/24 allow'], ['allow', 'deny'], true)
    EOS
  ) do |arguments|

    rescuable_exceptions = [ArgumentError]

    if defined?(IPAddr::InvalidAddressError)
      rescuable_exceptions << IPAddr::InvalidAddressError
    end

    raise Puppet::ParserError, 'validate_unbound_acl(): Wrong number of ' +
      "arguments given (#{arguments.size} for 3)" unless (1..3).include?(arguments.size)

    item     = arguments[0]
    actions  = arguments[1] || []
    extended = function_str2bool([arguments[2] || false])

    unless item.is_a?(Array)
      item = [item]
    end

    if item.size == 0
      raise Puppet::ParseError, 'validate_unbound_acl(): Requires an array ' +
        'with at least 1 element'
    end

    item.each do |i|
      unless i.is_a?(String)
        raise Puppet::ParseError, 'validate_unbound_acl(): Requires either ' +
          'an array or string to work with'
      end

      spec = i.split(/\s+/)

      case spec.size
      when 1
        # Just the address
        addr = spec[0]
        # If there is at least one action value then they're required
        if actions.size > 0
          raise Puppet::ParseError, 'validate_unbound_acl(): An action is ' +
            'expected'
        end
      when 2
        # An address followed by an action
        addr = spec[0]
        unless actions.include?(spec[1])
          raise Puppet::ParseError, 'validate_unbound_acl(): ' +
            "#{spec[1].inspect} is not a valid action"
        end
      else
        raise Puppet::ParseError, "#{i.inspect} is not a valid Unbound IP address."
      end

      # Match <address>@<port> syntax
      if addr =~ /^ ([^@]+) @ (\d+) $/x
        ip = $1
        unless $2.to_i < 65535
          raise Puppet::ParseError, "#{$2} is not a valid port."
        end
      else
        ip = addr
      end

      # Match extended IP address formats
      if extended
        case ip
        when /^ ([^\/]+) \/ (\d+) $/x
          # IP / Prefix
          begin
            unless IPAddr.new($1).mask($2).mask_addr.to_s(2).count('1') == $2.to_i
              raise Puppet::ParseError, "#{ip.inspect} is not a valid IP/Prefix pair."
            end
          rescue *rescuable_exceptions
            raise Puppet::ParseError, "#{ip.inspect} is not a valid IP/Prefix pair."
          end
        else
          # Normal IP
          begin
            x = IPAddr.new(ip)
            unless (x.ipv4? or x.ipv6?) and x.to_s.eql?(ip) and x.to_range.count == 1
              raise Puppet::ParseError, "#{ip.inspect} is not a valid IP address."
            end
          rescue *rescuable_exceptions
            raise Puppet::ParseError, "#{ip.inspect} is not a valid IP address."
          end
        end
      else
        # Normal IP
        begin
          x = IPAddr.new(ip)
          unless (x.ipv4? or x.ipv6?) and x.to_s.eql?(ip) and x.to_range.count == 1
            raise Puppet::ParseError, "#{ip.inspect} is not a valid IP address."
          end
        rescue *rescuable_exceptions
          raise Puppet::ParseError, "#{ip.inspect} is not a valid IP address."
        end
      end
    end
  end
end
